Add support for relative paths in debug prefix maps.

adrian-prantl opened the pull request:
Users may want to use -fdebug-prefix-map to remap a directory to "./".
This wasn't supported previously and resulted in duplicate path
prefixes in the debug info.

With this patch swiftc behaves similar to clang when remapping a path
to a relative path.

rdar://83753143

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves SR-NNNN.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@rmaz Let me know if you have any concerns about this.

keith commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -450,12 +450,17 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
+    StringRef RemappedFile(RemappedFileString);
     std::string CurDir = DebugPrefixMap.remapPath(Opts.DebugCompilationDir);
-    if (llvm::sys::path::is_absolute(RemappedFile)) {
+    // While Opts.DebugCompilationDir is guaranteed to be absolute,
```
Hrm is this true? I think there's a hole in this one today (that we utilize) where if you don't pass `-working-directory` this isn't set absolutely? Not sure

adrian-prantl replied:
Thanks for pointing that out. I reworded the comment.

adrian-prantl replied:
By the way, I didn't see that PR because I wasn't added a reviewer (and you probably couldn't add me due to how the repository is set up). Feel free to @-mention in PRs if you want a review.

adrian-prantl replied:
IIUC, this patch also fixes #39695 or are there other edge cases that I'm missing?

keith replied:
Thanks, I'm not entirely sure, but I think so, my test case is a bit different though with the nested directory case, can you add that same test case here?

adrian-prantl replied:
In my testcase the file is `%t/SubdirectoryWithAVeryUniqueName/debug_prefix_map_abs_rel.swift` and we remap `%t=.`.
In yours the file is `%/t/nested/dir/foo.swift` and we remap `%t/nested=.`. Since `%t` is the full path into the build directory, I don't see how the extra `nested` subdir makes a difference? Let me know if I'm missing something.

keith replied:
Hrm I thought there was a difference and it had something to do with `pwd` differing, but with some minor changes around the placement of `.` and `./` this change passes the tests from my PR

keith reviewed (commented):
Related https://github.com/apple/swift/pull/39695

adrian-prantl committed:
Add support for relative paths in debug prefix maps.

Users may want to use -fdebug-prefix-map to remap a directory to "./".
This wasn't supported previously and resulted in duplicate path
prefixes in the debug info.

With this patch swiftc behaves similar to clang when remapping a path
to a relative path.

rdar://83753143

adrian-prantl commented:
@swift-ci smoke test

rmaz commented:
> @rmaz Let me know if you have any concerns about this.

LGTM. This changes our `DW_AT_decl_file` entries from `./relative/folder/file.swift` to `relative/folder/file.swift` but otherwise I verified debugging still works correctly for a sample app.

adrian-prantl commented:
Fix test on Windows.

adrian-prantl commented:
@swift-ci smoke test

kastiglione commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -450,12 +450,18 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
+    SmallString<128> FileBuf(FileName);
+    llvm::sys::path::remove_dots(FileBuf);
+    std::string RemappedFileString = DebugPrefixMap.remapPath(FileBuf);
+    FileBuf.clear();
```
Instead of using `FileBuf` for two purposes (and needing this clear) what do you think of having separate vars?

adrian-prantl replied:
I wasn't sure if stack coloring was good enough detect that the memory can be reused and stack space comes at a premium in the compiler. That said this is pretty much a leaf function, so it probably wouldn't much.

kastiglione reviewed (commented):
After reading for a bit, I am not clear whether this subsumes https://github.com/apple/swift/pull/39695 or not. I do like that the equivalent patch for clang ([D111587](https://reviews.llvm.org/D111587)) has a test case where the `DIFile` `directory:` is a relative path other than `"."`.

kastiglione commented:
> This changes our `DW_AT_decl_file` entries from `./relative/folder/file.swift` to `relative/folder/file.swift`

Should we be explicit and product the former? When setting `target.source-map`, I have always seen the convention be `.=/some/path` and for me I like the idea of having an explicit `./` prefix in `DW_AT_decl_file`, since the `.` is present in both. Maybe this doesn't matter, but I wanted to bring it up.

rmaz commented:
> > This changes our `DW_AT_decl_file` entries from `./relative/folder/file.swift` to `relative/folder/file.swift`
> 
> Should we be explicit and product the former? Conventionally, I have always seen `settings set target.source-map . /some/path`. I like the idea of having a the debug info contain explicit `./` prefix in `DW_AT_decl_file`, since that way the `.` is present in the path being remapped, and the remapping. Maybe this doesn't matter, but I wanted to bring it up.

Thats actually a good point. I think we should match the paths output by clang, and there we have paths that start with the `./`.

adrian-prantl commented:
> > > This changes our `DW_AT_decl_file` entries from `./relative/folder/file.swift` to `relative/folder/file.swift`
> > 
> > 
> > Should we be explicit and product the former? Conventionally, I have always seen `settings set target.source-map . /some/path`. I like the idea of having a the debug info contain explicit `./` prefix in `DW_AT_decl_file`, since that way the `.` is present in the path being remapped, and the remapping. Maybe this doesn't matter, but I wanted to bring it up.
> 
> Thats actually a good point. I think we should match the paths output by clang, and there we have paths that start with the `./`.

@rmaz, can you help me reproduce what you are seeing? In my testcase the path in `DW_AT_decl_file` starts with `./` after remapping. What's different in your setup?

adrian-prantl commented:
This patch moves the call to remove_dots() to before remapping instead of after

rmaz commented:
> @rmaz, can you help me reproduce what you are seeing? In my testcase the path in `DW_AT_decl_file` starts with `./` after remapping. What's different in your setup?

You can test with the following:
```
% mkdir -p /tmp/cwd
% cd /tmp/cwd
% touch hello.swift
% swiftc -sdk $(xcrun --show-sdk-path) -g -c hello.swift -o hello.o -debug-prefix-map /tmp/cwd=.
% dwarfdump hello.o | grep DW_AT_decl_file
                  DW_AT_decl_file	("hello.swift")
```
Using the Xcode 13 swiftc gives:
```
                  DW_AT_decl_file	("./hello.swift")
```
Comparing with clang:
```
% echo "int main(){}" > hello.m
% clang -c hello.m -o hello.o -g -fdebug-prefix-map=/tmp/cwd=.
% dwarfdump hello.o | grep decl_file                          
                DW_AT_decl_file	("./hello.m")
```

adrian-prantl commented:
Thanks! 
I changed the algorithm to normalizing the path before remapping, and revert to only doing the common prefix stripping if file and curdir are absolute.

adrian-prantl commented:
@swift-ci smoke-test

adrian-prantl commented:
@swift-ci smoke test

adrian-prantl commented:
Test failure was unrelated.

adrian-prantl commented:
@swift-ci test and merge

keith commented:
@adrian-prantl can we cherry pick this to 5.5?

adrian-prantl commented:
Given how often we iterated to land on this solution, I think it's too risky to take this this late in the 5.5 branch's lifetime.
